lightning/src/ln/offers_tests.rs

```diff
@@ -727,6 +727,82 @@ fn creates_and_pays_for_offer_using_one_hop_blinded_path() {
 	expect_recent_payment!(bob, RecentPaymentDetails::Fulfilled, payment_id);
 }
 
+/// Verifies that the Invoice Request Message can be retried if it fails to reach the
+/// payee on the first attempt.
+#[test]
+fn creates_and_pays_for_offer_using_one_hop_blinded_path_with_retry() {
+	let chanmon_cfgs = create_chanmon_cfgs(2);
+	let node_cfgs = create_node_cfgs(2, &chanmon_cfgs);
+	let node_chanmgrs = create_node_chanmgrs(2, &node_cfgs, &[None, None]);
+	let nodes = create_network(2, &node_cfgs, &node_chanmgrs);
+
+	create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 10_000_000, 1_000_000_000);
+
+	let alice = &nodes[0];
+	let alice_id = alice.node.get_our_node_id();
+	let bob = &nodes[1];
+	let bob_id = bob.node.get_our_node_id();
+
+	let offer = alice.node
+		.create_offer_builder(None).unwrap()
+		.amount_msats(10_000_000)
+		.build().unwrap();
+	assert_ne!(offer.signing_pubkey(), Some(alice_id));
+	assert!(!offer.paths().is_empty());
+	for path in offer.paths() {
+		assert_eq!(path.introduction_node, IntroductionNode::NodeId(alice_id));
+	}
+	let payment_id = PaymentId([1; 32]);
+	bob.node.pay_for_offer(&offer, None, None, None, payment_id, Retry::Attempts(0), None).unwrap();
+	expect_recent_payment!(bob, RecentPaymentDetails::AwaitingInvoice, payment_id);
+
+	let _lost_onion_message = bob.onion_messenger.next_onion_message_for_peer(alice_id).unwrap();
+
+	// Simulate a scenario where the original onion_message is lost before reaching Alice.
+	// Use handle_message_received to regenerate the message.
+	bob.node.handle_message_received();
+	let onion_message = bob.onion_messenger.next_onion_message_for_peer(alice_id).unwrap();
+
+	alice.onion_messenger.handle_onion_message(&bob_id, &onion_message);
+
+	let (invoice_request, reply_path) = extract_invoice_request(alice, &onion_message);
+	let payment_context = PaymentContext::Bolt12Offer(Bolt12OfferContext {
+		offer_id: offer.id(),
+		invoice_request: InvoiceRequestFields {
+			payer_id: invoice_request.payer_id(),
+			quantity: None,
+			payer_note_truncated: None,
+		},
+	});
+	assert_eq!(invoice_request.amount_msats(), None);
+	assert_ne!(invoice_request.payer_id(), bob_id);
+	assert_eq!(reply_path.introduction_node, IntroductionNode::NodeId(bob_id));
+	let onion_message = alice.onion_messenger.next_onion_message_for_peer(bob_id).unwrap();
+	bob.onion_messenger.handle_onion_message(&alice_id, &onion_message);
+
+	// Expect no more OffersMessage to be enqueued by this point, even after calling
+	// handle_message_received.
+	bob.node.handle_message_received();
+
+	let result = bob.onion_messenger.next_onion_message_for_peer(alice_id);
+	match result {
```

Comment:
nit: combine this into one expression

Reply:
Done! [420ed56](https://github.com/lightningdevkit/rust-lightning/commit/420ed565bf1fce14e96c02e05fa06a7b49e3a3ec)